woocommerce/google-listings-and-ads
# Shipping Integration - Sync shipping rates for subdivisions of a country

### Changes proposed in this Pull Request:



Closes #1231.

This PR implements the system for parsing and syncing the suggested shipping rates for states/provinces and postal codes fetched from WC Shipping Zones (provided by #1341) with Google Merchant Center.

Two new adapter classes are created for mapping the store's shipping configuration to Google's `ShippingSettings`:

- **`DBShippingSettingsAdapter`**: Maps the rates obtained from the database to a ShippingSettings object that can be sent to the Google API. It accepts a `db_rates` argument that contains the rates in the same format as set in the `gla_shipping_rates` table.
- **`WCShippingSettingsAdapter`**: Maps the `LocationRate` objects fetched by parsing the WC Shipping Zone (by the `ShippingZone` class). It accepts a `rates_collections` parameter, which is an array of `CountryRatesCollection` containing location rates for each country supported by the store.

We group the shipping rates by their country and applicable shipping area first. An applicable area can be a whole country, a state, or specific postcodes of a country. Then for each group, we separate the rates that have any minimum order amount constraints (specifically free shipping rates with a min order amount threshold).

After the rates are all grouped together, we map them according to their applicable shipping area:

- If the rates only apply to a country, then we create a simple RateGroup with just one flat rate for the whole country.
- If the rates only apply to a state, then we create a table rate group and set the shipping rates for each state by mapping the rate to their associated location ID.
- If the rates apply to postcodes, we again create a table rate group for that country and set the rates for each group of postcodes.

An important note is that Google assumes all postal codes are unique within a country; therefore, the API doesn't accept different shipping rates associated with the same postcode but in different states/provinces. WooCommerce, on the other hand, allows the user to set different postcode values for any country+state combinations so there's a slight difference in how we handle that here. This shouldn't be a problem for us since we only support a handful of countries at the moment and AFAIK all those countries have unique postcodes. However, if we later add support for countries that don't adhere to this rule (like France) we (and Google) might need to change how this mapping works.

### Detailed test instructions:


1. Create multiple types of shipping settings, rates, and zones in WooCommerce settings
2. Edit your free listing program or go through the plugin onboarding and select the "Recommended" option for syncing the shipping settings
3. Complete setup and confirm in MC that all your shipping settings are synced (only the supported countries and methods will be synced as mentioned in #1228)
4. Edit your free listing program again and this time choose the "My shipping settings are simple." option
5. Enter shipping rates for the countries and complete setup
6. Check in MC that the shipping rates are now updated to reflect the new ones set in the previous step.


### Additional details:


### Changelog entry

> Add - Sync the shipping rates for states/provinces and postal codes to Merchant Center.
